### ISSUE TITLE
Standardise behaviour of params and optional arguments across parameterised attributes

### ISSUE THREAD

**Issue by ChrisMaddock** [nunit/nunit#2268]
NUnit has various parameterised arguments, which make various special conversions to adjust parameters to match method signatures, to allow invocation via reflection. The key ones I'm thinking of are `[TestCase()]`, `[TestCaseSource()]`, `[TestFixture()]` and `[TestFixtureSource()]`.

Support for params and optional arguments between these attributes is inconsistent. Over the years, this has raised various issues:
https://github.com/nunit/nunit/issues/1281 (open)
https://github.com/nunit/nunit/issues/1883 (open)
https://github.com/nunit/nunit/issues/1459 (open)
https://github.com/nunit/nunit/issues/1640
https://github.com/nunit/nunit/issues/1278
https://github.com/nunit/nunit/issues/1282
https://github.com/nunit/nunit/issues/257

We've talked in a couple of places ([link 1](https://github.com/nunit/nunit/pull/2104), [link 2](https://github.com/nunit/nunit/issues/1281), [link 3](https://github.com/nunit/nunit/pull/1850)) about harmonising this functionality, and having a single utility class, which resolves parameters passed to attributes, and ensures they are handled in a standard manner across all attributes. As I remember - this was something we agreed to go ahead with - @Yarmonov started on a PR in https://github.com/nunit/nunit/pull/2104. (This seems to have gone stale, unfortunately. Will check in on it shortly.) This would mean that the params keyword and optional arguments would be handled in a standard fashion across all of these attributes, meaning no unexpected surprises for the end-user. 

I wanted to create a master issue for this whole idea - as some of the existing open issues are tagged easyfix - and I didn't want anyone to end up implementing any more solutions individually, when we actually want to be sharing the existing functionality between attributes. If @nunit/framework-team agree - I'd like to close the open issues above, and use this as a single issue to reflect the overall plan.

**Comment by CharliePoole** [comment]
Seems like this should be an Epic.

**Comment by ChrisMaddock** [comment]
I would have thought an Epic implied we wanted people to take each issue on separately, where as my intention was that the functionality would be synchronized as one PR - rather than someone creating a new implementation for each attribute.

**Comment by Yarmonov** [comment]
I have done such utility class that covers all attributes you've mentioned, all cases (optional or/and array params), with a lot of performance optimizations, unit tests and some slight design ideas. Halfway migrated the existing code to this new utility before got extremely occupied. My apologizes.

Was going to find some time this weekend to finish this stuff.

In case I won't succeed, will push all code anyway. Hopefully you'll find it useful.

**Comment by ChrisMaddock** [comment]
Sounds great! 😃

I wasn't sure if you were still looking at this or not. If it's a work in progress, that's great. No rush - whenever you manage to fit it in. :-)

I've assigned this issue to you @Yarmonov. @nunit-framework-team - I'd still like to close the other issues, it sounds like they should hopefully be taken care of with @yarmonov's changes anyway.

**Comment by CharliePoole** [comment]
@ChrisMaddock Depends what you mean by hopeful. 😄 

If the others are the same issue, then I would close them as duplicates. If they are related and will only probably be fixed, then we should keep tracking and check them against the PR.

**Comment by Yarmonov** [comment]
I agree with @CharliePoole. There is a close to 100% probability that standardized solution will be better, but to be on the safe side it's better to ensure that results meet the expectations :) And then close tracked issues.

**Comment by ChrisMaddock** [comment]
Fair enough.

In that case, I'll just comment on the attached issues. What I don't want, is a new contributor unwittingly picking up one of those issues, when @Yarmonov has (probably!) already done the work for them. 😄

**Comment by pr-yemibedu** [comment]
Hello,
 Was some of the code of concepts from #1850 used here? That was closed but there did not seem to be any mention that directly linked the efforts here and on that PR. Thank you. Good day.

**Comment by CharliePoole** [comment]
I just ran into this WRT `TestCaseSoureAttribute` it turns out I cannot leave the final argument in `TestCaseData`. Looks like I have to provide an empty array, whereas I can just leave it out for `TestCaseAttribute`. 

This feels like a regression, but it may be a very old one - perhaps from day 1 of NUnit 3. 

Since I don't work on two things at the same time, I'm working around this for now.

**Comment by ChrisMaddock** [comment]
@nunit/framework-team If anyone's feeling particularly brave, I think this would be a great candidate for 4.0. It would allow us to standardise the behaviour of all these attributes and make any necessary breaking changes to achieve that.

**Comment by aaronburro** [comment]
Are there any plans to implement this? I see commits referencing this issue, but they don't seem to be supported yet (as of 3.13.2)

**Comment by stevenaw** [comment]
@ChrisMaddock I think v4 is a great opportunity to try and get this done. I can't remember if there were any concerns about breaking changes with this, but v4 would be the chance to address those if so.

@aaronburro There was some work done in the past but the PR ran stale before it could merge. I haven't checked, but if you're interested in helping add this to the framework the existing branch has likely done a decent chunk of the work and may just need some finishing off. I think any change here would likely need to target a release in v4.0

**Comment by CharliePoole** [comment]
@stevenaw Chris has withdrawn from working on NUnit. Someone else would need to take this on.

**Comment by stevenaw** [comment]
I see thanks @CharliePoole 
I wasn't suggesting Chris do it so much as responding to his comment. I'd have the background and interest to take this on myself (and I think I got partway there in the past) if time allowed in the future

**Comment by stevenaw** [comment]
@nunit/framework-team 
Many behaviours have been standardized in recent releases, including 4.5 which was released a few days ago.
I am going to close this under the assumption that future inconsistencies, if desired to be standardized, would have dedicated tickets for each behaviour
